sight_service/message_queue.py

```diff
@@ -0,0 +1,255 @@
+  ):
+    self.id_generator = id_generator
+    self.pending: Dict[int, T] = {}
+    self.active: Dict[str, Dict[int, T]] = {}
+    self.completed: Dict[int, T] = {}
+
+    self.batch_size = batch_size
+    self.pending_lock = lock_factory()
+    self.active_lock = lock_factory()
+    self.completed_lock = lock_factory()
+
+  def __str__(self) -> str:
+    all_messages = self.get_all_messages()
+
+    result = ['MessageQueue:']
+    result.append('  Pending Messages:')
+    for id, message in all_messages['pending'].items():
+      result.append(f'    ID: {id}, Message: {message}')
+
+    result.append('  Active Messages:')
+    for id, message in all_messages['active'].items():
+      result.append(f'    ID: {id}, Message: {message}')
+
+    result.append('  Completed Messages:')
+    for id, message in all_messages['completed'].items():
+      result.append(f'    ID: {id}, Message: {message}')
+
+    return '\n'.join(result)
+
+  # @overrides
+  def push_message(self, message: T) -> str:
+    """Pushes a message to the queue.
+
+    Args:
+      message: The message to push.
+    """
+    unique_id = self.id_generator.generate_id()
+    with self.pending_lock.gen_wlock():
+      self.pending[unique_id] = message
+    return unique_id
+
+  # @overrides
+  def process_messages(
+      self, worker_id: str, new_batch_size: Optional[int] = None
+    ) -> Dict[int, T]:
+    """Processes a batch of messages for a given worker.
+
+    Args:
+      worker_id: The ID of the worker that will process the messages.
+      new_batch_size: The size of the batch to process. If not provided, the
+        default batch size will be used.
+    """
+    batch_size = (
+        new_batch_size if new_batch_size is not None else self.batch_size
+    )
+    batch: Dict[int, T] = {}
+
+    with self.pending_lock.gen_wlock():
+      for _ in range(min(batch_size, len(self.pending))):
+        message_id = next(iter(self.pending))
+        message = self.pending.pop(message_id)
+        batch[message_id] = message
+
+    with self.active_lock.gen_wlock():
+      if worker_id not in self.active:
+        self.active[worker_id] = {}
+      self.active[worker_id].update(batch)
+
+    return batch
+
+  # @overrides
+  def complete_message(
+      self, message_id: str, worker_id: str, extra_details: Optional[any] = None
+    ) -> None:
+    """Completes a message of the given message ID of the given worker it moves it to the completed queue.
+
+    Args:
+      message_id: The ID of the message to complete.
+      worker_id: The ID of the worker that completed the message.
+      extra_details: Any extra details to add to the completed message. (only
+        for dict types messages)
+    """
+    with self.active_lock.gen_wlock():
+      if message_id in self.active.get(worker_id, {}):
+        message = copy.deepcopy(self.active[worker_id][message_id])
+        del self.active[worker_id][message_id]
```

Comment:
As I am deleting the same msg in the next line and msg can be a nested object or array and not only primitive data type to avoid loss of data , am copying msg before deleting

Reply:
Let me bring this one back up. You should not have to deepcopy to get this effect. By keeping a reference to the object in the message temporary var you're telling Python to not throw it away even when you delete this object from this.active. The way Python should do it it by counting the number of references to each memory object. If the number of references drops to 0 Python will garbage collect it. By keeping the reference via the temporary var you're telling Python that the memory object is being used.

Please try it without the deepcopy to confirm.

Reply:
Oh, that's fantastic !! , yes , I confirmed it , Python does keep the ref even after deleting it from the obj , Thanks !!